### PR TITLE
Added a helper to specifically create a LocalLocation from a URI

### PR DESCRIPTION
**Pull request by etorreborre**
@raronson

### CODE REVIEW

File: notion-core/src/main/scala/com/ambiata/notion/core/Location.scala [hunk 1]
```diff
@@ -91,6 +91,41 @@ object Location {
       e.getMessage.left
   }
 
+  def localLocationFromUri(s: String): String \/ Location =
+    try {
+      val uri = new java.net.URI(s)
+      uri.getScheme match {
+        case "file" =>
+          LocalLocation(uri.toURL.getFile).right
```
Comment (raronson) [inline notion-core/src/main/scala/com/ambiata/notion/core/Location.scala]: Why do you need to convert to `URL`? doesn't `getPath` do the same?

File: notion-core/src/main/scala/com/ambiata/notion/core/Location.scala [hunk 2]
```diff
@@ -91,6 +91,41 @@ object Location {
...
+        case null =>
+          LocalLocation(uri.getPath).right
+        case _ =>
+          s"Not a local location [${uri.getScheme}]".left
+      }
+    } catch { case e: java.net.URISyntaxException => e.getMessage.left }
```
Comment (raronson) [inline notion-core/src/main/scala/com/ambiata/notion/core/Location.scala]: If there are other exceptions like `NullPointerException` you can catch it with `\/.fromTryCatchNonFatal`

File: notion-core/src/main/scala/com/ambiata/notion/core/Location.scala [hunk 3]
```diff
@@ -91,6 +91,41 @@ object Location {
...
   }
 
+  def localLocationFromUri(s: String): String \/ Location =
+    try {
+      val uri = new java.net.URI(s)
+      uri.getScheme match {
+        case "file" =>
+          LocalLocation(uri.toURL.getFile).right
+        case null =>
+          LocalLocation(uri.getPath).right
```
Comment (raronson) [inline notion-core/src/main/scala/com/ambiata/notion/core/Location.scala]: Should we be defaulting to local location when a scheme isn't present?

File: notion-core/src/main/scala/com/ambiata/notion/core/Location.scala [hunk 4]
```diff
@@ -91,6 +91,41 @@ object Location {
...
+          s"Not a local location [${uri.getScheme}]".left
+      }
+    } catch { case e: java.net.URISyntaxException => e.getMessage.left }
+
+  def s3LocationFromUri(s: String): String \/ Location =
+    try {
+      val uri = new java.net.URI(s)
+      uri.getScheme match {
+        case "s3" =>
+          S3Location(uri.getHost, uri.getPath.drop(1)).right
```
Comment (raronson) [inline notion-core/src/main/scala/com/ambiata/notion/core/Location.scala]: The URI `s3:/bucket/key` will mean `scheme` will be `s3` but `getHost` will return `null`

**Comment by raronson** [comment]
I think there needs to be more guards/checks when using `URI` for parsing these. eg. the following will have the `s3` scheme, but calling `getHost` and `getPath` has different behaviors:

  * `s3:foo/bar` - host `null`, path `null`
  * `s3:/foo/bar` - host `null`, path `/hello/world`
  * `s3://foo/bar` - host `foo`, path `/bar`
  * `s3:///foo/bar` - host `null`, path `/foo/bar`

**Comment by etorreborre** [comment]
@raronson my aim for now was just to check if something is really a local location without touching the existing behaviour. The fixup commit makes this clearer.

**Comment by raronson** [comment]
@etorreborre your aim looks like its captured correctly, but we should revisit the original parsing code at some point as it has the potential to cause problems

**Pull request merged**